### PR TITLE
[adding] ADR-21 NATS Configuration Contexts

### PR DESCRIPTION
**Pull request by ripienaar**
Signed-off-by: R.I.Pienaar <rip@devco.net>

### CODE REVIEW

**Review by ColinSullivan1 — COMMENTED** [review]
Let's discuss Q1 22. It's a great feature, we just need to determine where it'll be implemented.

**Comment by ripienaar** [comment]
> Let's discuss Q1 22. It's a great feature, we just need to determine where it'll be implemented.

👍 for now this is just capturing the feature, not communicating an intend for clients to implement.

**Review by aricart — APPROVED** [review]
LGTM

**Review by aricart — APPROVED** [review]
LGTM

File: adr/ADR-21.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+|`ca`| |Path to the x509 Certificate Authority|
+|`nsc`| |A `nsc` resolve url for loading credentials and server urls|
+|`jetstream_domain`| |The JetStream Domain to use|
+|`jetstream_api_prefix`| |The JetStream API Prefix to use|
+|`jetstream_event_prefix`| |The JetStream Event Prefix|
+|`inbox_prefix`| |A prefix to use when generating inboxes|
+
+All fields are optional, none are marked as `omitempty`, users wishing to edit these with an editor should known all valid key names.
+
+Above settings map quite obviously to client features with the exception of `nsc`, the `nsc` key takes a URL like value, examples are:
```
Comment (matthiashanel) [inline adr/ADR-21.md:71]: Should that nsc be invoked with store and nkey directory specified by the context as well?

Comment (ripienaar) [reply adr/ADR-21.md:71]: Quite happy to let those who choose to customise everything carry that  complexity burden themselves by just pointing at the credential file themselves. 

This infinite customizability we are adding at every level is a plague. 

Why would nsc need a context? @aricart and I discussed that long ago and decided it’s not compatible models

Comment (ripienaar) [reply adr/ADR-21.md:71]: Also, afaik, nsc reads environment variables and a config file, not options you pass? `NSC_HOME` and `NKEYS_PATH` etc, I am not going to be supporting passin these around. If someone sets them they will continue to work for nsc when invoked by the context.

**Comment by tbeets** [comment]
In sample usage API, the value of OS environment variable would be the filename of the desired context or the friendly name of the desired context?   Both options would seem useful.   @ripienaar

**Comment by ripienaar** [comment]
> In sample usage API, the value of OS environment variable would be the filename of the desired context or the friendly name of the desired context? Both options would seem useful. @ripienaar

No file name based access - its just the context name. In future we'd have non file based context storage, so its only name.

**Comment by tbeets** [comment]
LGTM.  I think whether we factor in more NSC stuff entirely related to futures of CLI tools (i.e. what and how many).

**Comment by derekcollison** [comment]
I think this direction and exposing this to apps and scripts and automation is spot on.

One observation I have had is that a bunch of popular tools just own `~/.<proj>`, so for us would be `~/.nats` maybe.

**Comment by ripienaar** [comment]
> One observation I have had is that a bunch of popular tools just own `~/.<proj>`, so for us would be `~/.nats` maybe.

That's the nice thing about standards, we can just follow the it without thinking what some others tools or dont. The standard covers everything from local data storage, binaries and more - storage paths for example that we need for things like `nats server run` and integrates with the eco system at large.

Why would we choose something that specifically would work against the desktop os direction?

**Comment by derekcollison** [comment]
My $0.02 is that default standards outweigh the formalized ones, who continually seem to be losing power. So wanted to see what the popular projects with lots of users were doing. I will check if anyone uses that convention today or if they have noted they would switch, but I am not aware of any. Are you?

**Comment by ripienaar** [comment]
It has VERY WIDE adoption. Few kind of small projects like Emacs, curl, git, GTK, helm, ImageMagick, less and a pile of standard shell things, every linux desktop environment, systemd and its dynamically mounted home directories, mercurial, mutt, nano, neovim, python pip, tmux, etc.

**Comment by derekcollison** [comment]
Good to know. Been using emacs forever so it is still `~/.emacs` for me. A bunch of the other ones as well, but maybe new installs default to ``~/.config/<project>`

**Comment by ripienaar** [comment]
Yes, obviously backwards compat on tool like the listed ones is fairly important, but the gravity is towards this. Emacs checks `~/.config/emacs` in adition to older paths.

**Comment by tbeets** [comment]
I see about 50/50 sticking their stuff under ~/.config by default.   JetBrains, Google Chrome, and NATS CLI are three examples that do.   Rust, Mozilla, Maven, minikube, AWS, and Azure are examples of those that splat in the homedir.    Lots of variability.

**Comment by tbeets** [comment]
Definitely support sticking application-specific stuff under `~/.config` to separate from typical shell config etc. in homedir if we have to make a default location call either way.

**Comment by ripienaar** [comment]
Arch wiki has a partial list of things that supports it https://wiki.archlinux.org/title/XDG_Base_Directory#Supported

**Review by jnmoyne — APPROVED** [review]
LLANBTM

(Looks like a no brainer to me)

**Pull request merged**